# 红宝石凸起

> 原文：<https://www.educba.com/ruby-raise/>

![Ruby raise](img/9116ff6162906bc1a60a7d142757a62b.png)



## Ruby raise 简介

很多时候，当我们在处理一组非常困难场景时，代码很可能会执行失败(比如打开一个文件)，例如，如果我们要在任何文件上写一些内容，而该文件不存在，在这种情况下，代码执行将会停止，这是不正确的，我们需要处理它。在 Ruby 中我们有一种方法来处理这些情况，那就是抛出语句。raise 是 Ruby 中的一个关键字，它允许我们在发现异常时抛出异常，raise 将抛出一个异常，这个异常将在 rescue 语句中被捕获。

**语法:**

<small>网页开发、编程语言、软件测试&其他</small>

下面是 Ruby 中 raise 语句的简单语法。在下面的语法中，有两个最重要的活动正在进行；它们在下面给出。

*   首先，我们提出一个例外。
*   在第二个活动中，我们使用关键字 rescue 进行救援。

`begin
raise
# Inside the raise block the exception will be raise
rescue
# The raise exception will be captured here to handle.
end`

### Ruby 中的 raise 语句是如何工作的？

在深入解释工作之前，我们应该知道应该在哪里使用 raise 语句，在打开任何文件和处理数据库相关查询的情况下，我们可以将 raise 与 rescue 一起使用。因为在这些情况下，我们无法预测情况的本质。文件可以在那里，也可以不在那里，同样，我们为任何表编写的查询也可能不在数据库中，这可能会停止代码的进一步执行。

*   在这里，我们在 begin 块中编写的每个代码都受到保护，这意味着从开始到救援的所有代码都处于安全模式(任何异常都将在那里处理)。如果我们试图打开任何文件或连接到数据库，并且发生任何异常，在这种情况下，raise 将引发一个问题。
*   如果在 begin 块中发生异常，它将停止，控制将在 rescue 和 end 之间转移。如果从 raise 块中出现异常，我们可以为用户返回一个有效的消息，从这个消息中，我们可以很容易地理解异常的类型。
*   每次执行 rescue 语句时，Ruby 都会检查并比较每个参数引发的异常。

### 在 Ruby 中实现 raise 的示例

下面是 Ruby 中 raise 的例子:

#### 示例#1

在下面的例子中，我们使用了一个非常简单的方法来理解 Ruby 中的筹资流程。我们使用了 begin 语句。我们已经完成了 begin 语句中的筹集和救援部分。在下面的例子中，有两个最重要的活动正在进行；它们在下面给出。

*   首先，我们引发了一个异常，在这里我们可以引发它，以防它发现任何代码在继续执行时停止执行的情况。因此，它只会引发一个问题，并停止进一步的执行。
*   在第二个活动中，我们使用关键字 rescue 进行救援。它将捕捉是否发现任何引发的语句。

请看下面的例子。

**代码:**

`begin
# Here we are raising an issue
raise 'Here we have raised an issue'
puts 'I am after the raising an issue'
# rescue for the issue
rescue
puts 'Capturing the raised issue here'
end`

**输出:**

![Ruby raise Example 1](img/ba8b5b695457ef7d48e766e456280fc8.png)



#### 实施例 2

在下面的例子中，我们使用了一个简单的 raise 语句流。让我解释下面的例子和一些步骤。在 Ruby 中使用 raise 语句是一种函数式方法。

*   首先，我们创建了一个方法 raise_example 方法。
*   我们在方法的末尾调用方法 raise_example。
*   在方法 raise_example 中，我们做了两件事，一件是引发一个异常，另一件是在 rescue 语句中引发问题。
*   raise 和 rescue 这两个工作我们都写在了 begin 块中，这是 Ruby 中处理 raise 语句的一个非常安全的地方。

请看下面的例子。

**代码:**

`def raise_example
begin
puts 'This is going to get called before the raise statement!'
# using raise to create an exception   which will get capture in rescue
raise 'Here we are creating an exception with raising it'
puts 'This is after raise statement'
# Using Rescue method
rescue
puts 'Now we are inside the rescue statement , it will rescue the raise issue'
end
puts 'This is the end of begin'
end
# calling method raise_example
raise_example`

**输出:**

![raisemethod Example 2](img/7556e7be9078aa316768e08d88e56549.png)



#### 实施例 3

在这里，我们将 assure 与 raise 语句一起使用，sequence 将确认问题是否被提出。我们可以在下面的步骤中解释 raise 语句。

*   它的工作方式几乎类似于简单的 try-catch，唯一不同的是它比 else 多了一些特性，确保处理异常。在这里，我们将确保语句与提高语句一起使用。
*   我们可以看到下面的例子，这里我们拯救了一个在 begin 块中引发的异常，rescue 语句将获取数据消息，这也可以是任何自定义消息，我们将在 rescue 中获得由 raise 语句引发的消息。
*   最后，我们还使用了确保该问题的确保语句。

请遵循下面的例子。

**代码:**

`begin
puts 'Searching for the file abc.txt,print success if the file is available'
raise 'Searching for the file klm.txt,print the fail if file is not there'
rescue Exception => exception
puts exception.message
puts exception.backtrace.inspect
ensure
puts "We are able to ensure that it was raise issue we need to take care of it"
end`

**输出:**

![Ruby raise Example 3](img/0decca6eadf9407d45228536a27de2ad.png)



### 结论

从本教程中，我们学习了 Ruby 中 raise 语句的流程，了解了 Ruby 的基本语法，了解了可以使用 raise 语句的地方，还了解了 raise 语句的工作流程。

### 推荐文章

这是一个红宝石饲养指南。在这里，我们将借助一些有用的例子和代码实现来讨论介绍以及如何在 Ruby 中使用语句。您也可以浏览我们推荐的其他文章，了解更多信息——

1.  [Ruby 命令](https://www.educba.com/ruby-commands/)
2.  [红宝石的用途](https://www.educba.com/uses-of-ruby/)
3.  [Perl vs Ruby](https://www.educba.com/perl-vs-ruby/)





